### PR TITLE
Fix handling of filename for http file

### LINKED ISSUE #3532: [Bug][http] can't declare requiredness for meta properties about  file part of multipart
According to https://github.com/microsoft/typespec/issues/3046, `httpFile` could be used to change the requiredness of the optional metadata properties

image

However, when I try the following code, I find the type is Scalar bytes which contains no info about "contentType"/"filename" is required for this file part.
```
  it("test", async () => {
    const op = await getHttpOp(`

    @Private.httpFile
    model RequiredFileMetadata {
      contentType: string;
      filename: string;
      contents: bytes;
    }
      op read(
        @header contentType: "multipart/mixed",
        @multipartBody body: [
          HttpPart<RequiredFileMetadata, #{ name: "file" }>
        ]): void;
    `);

    const body = op.parameters.body;
    strictEqual(body?.bodyKind, "multipart");
  });
```
image

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #3532

### CODE REVIEW

File: packages/http/test/multipart.test.ts [hunk 1]
```diff
@@ -153,6 +153,56 @@ describe("resolving part payload", () => {
...
+          @header contentType: "multipart/mixed",
+          @multipartBody body: [
+            HttpPart<FileWithFilename, #{ name: "file" }>
+          ]): void;
+        `);
+
+      const body = op.parameters.body;
+      strictEqual(body?.bodyKind, "multipart");
+      strictEqual(body.parts.length, 1);
+      ok(body.parts[0].filename, "Filename property should have been resolved");
```
Comment (msyyc) [inline packages/http/test/multipart.test.ts:201]: With this PR, author could set filename as required or not explicitly. What about "content-type"? If author also wants to set `content-type` of this httppart as required, current API can't show the info.

Comment (timotheeguerin) [reply packages/http/test/multipart.test.ts:201]: Content type is already exposed in the body as content type

Comment (timotheeguerin) [reply packages/http/test/multipart.test.ts:201]: File name is the only special datat that’s not part of the body

Comment (msyyc) [reply packages/http/test/multipart.test.ts:201]: According to https://datatracker.ietf.org/doc/html/rfc7578#section-4.4, content-type is not `MUST` but `MAY` or `SHOULD` which means that it is not necessary part in payload. However, now Typespec author could use `httpFile` to mark `filename` and `content-type` as required, so Typespce compiler shall expose the requiredness info for language emitters. Currently https://github.com/microsoft/typespec/blob/e1bf4607ede6e0de4896607e80440b70e3363e0b/packages/http/src/types.ts#L445-L448, contentTypes of body is `string[]` and may have default value set by Typespec compiler which language emitter can't infer the requiredness for it.
nit: Java/.NET may need it so add @chunyu3 and @weidongxu-msft
image

**Comment by weidongxu-microsoft** [comment]
Is there a way to say "filename default to <some_name> if not provided by user" (a kind of client default)?

File: packages/http/test/multipart.test.ts (hunk 1 above)
Comment (weidongxu-microsoft) [reply packages/http/test/multipart.test.ts:201]: There is `headers`? Does `content-type` header belongs to it?
https://github.com/microsoft/typespec/blob/e1bf4607ede6e0de4896607e80440b70e3363e0b/packages/http/src/types.ts#L493-L494

Comment (msyyc) [reply packages/http/test/multipart.test.ts:201]: > There is `headers` in `HttpOperationPart`? Does `content-type` header belongs to it?
> 
> https://github.com/microsoft/typespec/blob/e1bf4607ede6e0de4896607e80440b70e3363e0b/packages/http/src/types.ts#L493-L494

`headers` here contains headers except `content-type`

**Comment by msyyc** [comment]
> Is there a way to say "filename default to <some_name> if not provided by user" (a kind of client default)? -- the usual case for this is that filename is required on backend and client want to provide a default (e.g. openai)

Ever discussed here https://github.com/microsoft/typespec/issues/3046#issuecomment-2105434345 while it seems no plan to support "define default value in client side"

File: packages/http/test/multipart.test.ts (hunk 1 above)
Comment (chunyu3) [reply packages/http/test/multipart.test.ts:201]: For meta-data like `filename` and `content-type` of each part, if customer define this those meta-data in spec, .NET will expose this meta-data as parameter or property to provide customer the capability to set. If not, .NET will not expose it. So for .net, we need to know if the `content-type` is defined or not as `filename`

**Comment by chunyu3** [comment]
> > Is there a way to say "filename default to <some_name> if not provided by user" (a kind of client default)? -- the usual case for this is that filename is required on backend and client want to provide a default (e.g. openai)
> 
> Ever discussed here [#3046 (comment)](https://github.com/microsoft/typespec/issues/3046#issuecomment-2105434345) while it seems no plan to support "define default value in client side" in Typespec.

.NET does not honor `client-side-default-value`, but we honor constant. so if the `filename` is define as `filename: "file"` it is a constant, .NET will not expose parameter for `filename`, but will set `filename=file` in the http request in content-dispose header.

File: packages/http/test/multipart.test.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/http/test/multipart.test.ts:201]: Added `contentTypeProperty` to the `HttpOperationBody`

File: packages/http/src/types.ts [hunk 2]
```diff
@@ -445,6 +445,8 @@ export interface HttpOperationResponseContent {
 export interface HttpOperationBodyBase {
   /** Content types. */
   readonly contentTypes: string[];
+  /** Property used to set the content type if exists */
+  readonly contentTypeProperty?: ModelProperty;
```
Comment (msyyc) [inline packages/http/src/types.ts:449]: Like `filename`, this property shall be in `HttpOperationPart` instead of `HttpOperationBodyBase`

**Pull request merged**